packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,190 @@
+export class SchemaReporter {
+  // These mirror the gql variables
+  private readonly serverInfo: EdgeServerInfo;
+  private readonly executableSchemaDocument: any;
+  private readonly url: string;
+
+  private isStopped: boolean;
+  private readonly headers: Headers;
+
+  constructor(
+    serverInfo: EdgeServerInfo,
+    schemaSdl: string,
+    apiKey: string,
+    schemaReportingEndpoint: string | undefined,
+  ) {
+    this.headers = new Headers();
+    this.headers.set('Content-Type', 'application/json');
+    this.headers.set('x-api-key', apiKey);
+    this.headers.set('apollographql-client-name', 'apollo-engine-reporting');
+    this.headers.set(
+      'apollographql-client-version',
+      require('../package.json').version,
+    );
+
+    this.url =
+      schemaReportingEndpoint ||
+      'https://engine-graphql.apollographql.com/api/graphql';
+
+    this.serverInfo = serverInfo;
+    this.executableSchemaDocument = schemaSdl;
+    this.isStopped = false;
+  }
+
+  public stopped(): Boolean {
+    return this.isStopped;
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  public async reportServerInfo(
+    withExecutableSchema: boolean,
+  ): Promise<ReportServerInfoReturnVal> {
+    const { data, errors } = await this.graphManagerQuery({
+      info: this.serverInfo,
+      executableSchema: withExecutableSchema
+        ? this.executableSchemaDocument
+        : null,
+    });
+
+    if (errors) {
+      throw new Error((errors || []).map((x: any) => x.message).join('\n'));
+    }
+
+    function msgForUnexpectedResponse(data: any): string {
+      return [
+        'Unexpected response shape from Apollo Graph Manager when',
+        'reporting server information for schema reporting. If',
+        'this continues, please reach out to support@apollographql.com.',
+        'Received response:',
+        JSON.stringify(data),
+      ].join(' ');
+    }
+
+    if (!data || !data.me || !data.me.__typename) {
+      throw new Error(msgForUnexpectedResponse(data));
+    }
+
+    if (data.me.__typename === 'UserMutation') {
+      this.isStopped = true;
+      throw new Error(
+        [
+          'This server was configured with an API key for a user.',
+          "Only a service's API key may be used for schema reporting.",
+          'Please visit the settings for this graph at',
+          'https://engine.apollographql.com/ to obtain an API key for a service.',
+        ].join(' '),
+      );
+    } else if (data.me.__typename === 'ServiceMutation') {
+      if (!data.me.reportServerInfo) {
+        throw new Error(msgForUnexpectedResponse(data));
+      }
+      return data.me.reportServerInfo;
+    } else {
+      throw new Error(msgForUnexpectedResponse(data));
+    }
+  }
+
+  private async graphManagerQuery(
+    variables: ReportServerInfoVariables,
+  ): Promise<SchemaReportingServerInfoResult> {
+    const request: GraphQLRequest = {
+      query: reportServerInfoGql,
+      operationName: 'ReportServerInfo',
```

Comment:
There's no need to include this unless you have more than one (named) operation in `query`.